### PR TITLE
Backfill `:effective-type` for coerced fields

### LINKED ISSUE #42931: When Filtering with Dates Coerced from Numeric Fields in a Table Visualization - Filter Sometimes Disappears
### Describe the bug

When a numeric value is coerced to a date there is some unexpected behavior when you use the date as a filter on a table visualization. the filter works, but under certain conditions it will disappear while you're interacting with other filters.

### To Reproduce

1. Coerce a numeric field to a date (you can test this with the ID field on any of the sample DB tables)
2. Create a simple question and display results in a table viz
3. Filter any field on the table
4. Filter by the coerced date field
5. Edit the first field filter
6. Then one of two things may happen:
- If there were records that matched the date filter, the filter does not disappear
- If there were not records that matched the coerced date filter then it vanishes

### Expected behavior

It should mirror the behavior of a date filter that uses a non coerced date field. The field filter shouldn't vanish when the other fields are edited.

### Logs

_No response_

### Information about your Metabase installation

```JSON
{
  "browser-info": {
    "language": "en-US",
    "platform": "Win32",
    "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.23+9",
    "java.vendor": "Eclipse Adoptium",
    "java.vendor.url": "https://adoptium.net/",
    "java.version": "11.0.23",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.23+9",
    "os.name": "Linux",
    "os.version": "5.15.133.1-microsoft-standard-WSL2",
    "user.language": "en",
    "user.timezone": "GMT"
  },
  "metabase-info": {
    "databases": [
      "h2",
      "postgres"
    ],
    "hosting-env": "unknown",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "11.22 (Debian 11.22-1.pgdg110+1)"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.7.2"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2024-05-09",
      "tag": "v1.49.9",
      "hash": "c0913c7"
    },
    "settings": {
      "report-timezone": null
    }
  }
}
```

### Severity

Annoying

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by lbrdnk**
Closes #42931

This PR adds functionality that backfills `:effective-type` for coerced fields.

Apart from disappearing filter, mentioned in the linked issue, there was another problem occurring on master only. Query execution was completing with error. Reason was same as for the filter disappearance -- coerced fields contained `:base-type` only (no `:effective-type`), hence query containing those did not pass schema validation.

### CODE REVIEW

File: src/metabase/lib/query.cljc [hunk 1]
```diff
@@ -166,31 +187,31 @@
...
-                                         (m/remove-vals
-                                           #(= :type/* %)
-                                           (-> (lib.expression/expression-ref query stage-number expression-name)
-                                               second
-                                               (select-keys [:base-type :effective-type])))
-                                         (catch #?(:clj Exception :cljs :default) _
-                                           ;; This currently does not find expressions defined in join stages
-                                           nil))]
-                         ;; Fallback if metadata is missing
-                         [:expression (merge found-ref opts) expression-name]))))
```
Comment (lbrdnk) [inline src/metabase/lib/query.cljc:193]: Editor having its whitespace opinions.

File: e2e/test/scenarios/binning/reproductions/23851-drill-temporal-extraction.cy.spec.js [hunk 2]
```diff
@@ -48,12 +48,12 @@ describe("issue 23851", () => {
...
     cy.findByTestId("filter-pill").should(
       "have.text",
-      "Created At: Day of week is equal to 6",
+      "Created At: Day of week is on Friday",
     );
     cy.get("[data-testid=cell-data]").should("contain", "109.22");
     cy.icon("notebook").click();
     getNotebookStep("filter")
-      .findByText("Created At: Day of week is equal to 6")
+      .findByText("Created At: Day of week is on Friday")
```
Comment (lbrdnk) [inline e2e/test/scenarios/binning/reproductions/23851-drill-temporal-extraction.cy.spec.js:56]: I think more appropriate phrasing would be "Day of week is Friday". I'll check for the ways how to change that.

File: src/metabase/query_processor/util.clj [hunk 3]
```diff
@@ -114,9 +115,13 @@
...
-                (cond-> query
-                  (#{"query" "native"} (:type query)) (#(lib.convert/->pMBQL (mbql.normalize/normalize %)))
-                  (#{:query :native} (:type query))   lib.convert/->pMBQL)
+                ;; Expression type check supression is necessary because coerced fields in `query` may not have
+                ;; `:effective-type` populated. That's the case during call to this function in
+                ;; `process-userland-query-middleware` that occurs before normalization.
+                (binding [lib.schema.expression/*suppress-expression-type-check?* true]
+                  (cond-> query
+                    (#{"query" "native"} (:type query)) (#(lib.convert/->pMBQL (mbql.normalize/normalize %)))
+                    (#{:query :native} (:type query))   lib.convert/->pMBQL))
```
Comment (lbrdnk) [inline src/metabase/query_processor/util.clj:124]: This suppresses BE warning, but looking back, I believe the warning should not be suppressed. We should rather ensure that `->pMBQL` is called on correct queries (eg. not containing filter expressions that use datetime functions on coerced fields containing only base type and no effective type).

Comment (lbrdnk) [reply src/metabase/query_processor/util.clj:124]: I'd like to consult this on weekly before merging.

File: e2e/test/scenarios/binning/reproductions/23851-drill-temporal-extraction.cy.spec.js (hunk 2 above)
Comment (lbrdnk) [reply e2e/test/scenarios/binning/reproductions/23851-drill-temporal-extraction.cy.spec.js]: Resolved by modification to `lib.metadata.calculation/display-name-method ::varargs`

File: src/metabase/lib/remove_replace.cljc [hunk 4]
```diff
@@ -377,15 +384,16 @@
     stage-number :- :int
     target-clause
     new-clause]
-   (cond
-     (and (map? target-clause) (= (:lib/type target-clause) :mbql/join))
-     (replace-join query stage-number target-clause new-clause)
+   (let [query (backfill-effective-type query)]
```
Comment (snoe) [inline src/metabase/lib/remove_replace.cljc:387]: Why is this necessary here? How do you get a query (built through `query-method`) without effective-type now?

File: src/metabase/lib/query.cljc [hunk 5]
```diff
@@ -106,6 +106,36 @@
                        (not (lib.util.match/match-one (:expressions stage) :offset)))))
                (:stages query))))
 
+(defn add-effective-type-to-coereced-fields
+  "Add `:base-type` and `:effective-type` (where applicable) to options of all fields in `x`, using `metadata-provider.
```
Comment (snoe) [inline src/metabase/lib/query.cljc:110]: I don't see base-type being added here (but I think it should be).

File: src/metabase/lib/query.cljc [hunk 6]
```diff
@@ -171,14 +202,14 @@
               (map (fn [[stage-number stage]]
                      (lib.util.match/replace stage
                        [:field
-                        (opts :guard (every-pred map? (complement (some-fn :base-type :effective-type))))
+                        (opts :guard (every-pred map? (complement (every-pred :base-type :effective-type))))
```
Comment (snoe) [inline src/metabase/lib/query.cljc:205]: I believe this `(map... )`should be calling `add-effective-type-to-coereced-fields` and merging the work between the two (expression types and `base-type` from here, move to the fn)

File: test/metabase/lib/remove_replace_test.cljc [hunk 7]
```diff
@@ -1423,3 +1426,46 @@
...
+                           :filters      [old-clause
+                                          missing-effective-type-clause]}]}
+        new-clause [:=
+                    {:lib/uuid "89c7eba5-d0a8-4a45-b0bf-ea44de20ee70"}
+                    [:field
+                     {:lib/uuid  "cfadec82-7727-4901-a557-5dca8530d8d9",
+                      :base-type :type/Text}
+                     (meta/id :people :city)]
+                    "Aberdeen"]]
+    (mu/disable-enforcement
```
Comment (snoe) [inline test/metabase/lib/remove_replace_test.cljc:1466]: It doesn't seem right that enforcement needs to be disabled here. Why is this needed?

File: test/metabase/lib/remove_replace_test.cljc [hunk 8]
```diff
@@ -1423,3 +1426,46 @@
...
                                 (lib/min (lib/get-month (by-name (lib/orderable-columns join-query 0) "CREATED_AT"))))))))
+
+(deftest ^:parallel coerced-filter-test
+  (let [effective-type (types/effective-type-for-coercion :Coercion/UNIXSeconds->DateTime)
+        mp (merged-mock/merged-mock-metadata-provider
+            meta/metadata-provider
+            {:fields [{:id                (meta/id :people :id)
+                       :coercion-strategy :Coercion/UNIXSeconds->DateTime
+                       :effective-type    effective-type}]})
+        old-clause [:=
```
Comment (snoe) [inline test/metabase/lib/remove_replace_test.cljc:1437]: You should build clauses and queries with `lib` functions
```
(-> (lib/query meta/metadata-provider (meta/table-metadata :people)
  (lib/filter (lib/= (meta/field-metadata :people :city) "Abilene")))
```


and use `lib/filters` to get a reference to `old-clause`

If you need to drop `effective-type` from a clause use an `update-in [:filters 1 ...]` to drop it from the query.

**Review by snoe — CHANGES_REQUESTED** [review]

File: src/metabase/lib/remove_replace.cljc (hunk 4 above)
Comment (lbrdnk) [reply src/metabase/lib/remove_replace.cljc:387]: Oh, it is not necessary anymore. Fixing `query-method :mbql/query` did the trick. Thanks.

File: src/metabase/lib/query.cljc (hunk 5 above)
Comment (lbrdnk) [reply src/metabase/lib/query.cljc:110]: Function was updated to handle also base types.

File: src/metabase/lib/query.cljc (hunk 6 above)
Comment (lbrdnk) [reply src/metabase/lib/query.cljc:205]: Now `add-types-to-fields` is called here and expression types are added to its result.

File: test/metabase/lib/remove_replace_test.cljc (hunk 7 above)
Comment (lbrdnk) [reply test/metabase/lib/remove_replace_test.cljc:1466]: That test is redundant now as changes to the `replace-clause`. `disable-enforcement` was used so replace query can be called on malformed query. That was happening on FE before change to `query-method :mbql/query`. I've found this as simplest way of making that test compatible with clj and cljs. I should have added comment about that.

File: test/metabase/lib/remove_replace_test.cljc (hunk 8 above)
Comment (lbrdnk) [reply test/metabase/lib/remove_replace_test.cljc:1437]: Thanks, I'll use lib functions to build queries in tests going forward.

File: src/metabase/lib/query.cljc [hunk 9]
```diff
@@ -141,7 +182,8 @@
 
 (defmethod query-method :query ; legacy MBQL query
   [metadata-providerable legacy-query]
-  (query-from-legacy-query metadata-providerable legacy-query))
+  (query-from-legacy-query metadata-providerable
```
Comment (snoe) [inline src/metabase/lib/query.cljc:185]: if these calls were inverted `(add-types-to-fields (query-from-legacy-query ....))` would that make handling legacy fields unnecessary in `add-types-to-fields`? why not?

Comment (lbrdnk) [reply src/metabase/lib/query.cljc]: Problem with that approach is that `query-from-legacy-query` calls `query-with-stages`. `query-with-stages` must return `::lib.schema/query` to conform schema. Hence the reversion of calls produces the error. That error was the other problem, present on master only.

To simplify `add-types-to-fields`, I've moved the call to it into `query-from-legacy-query` right after the query is transformed to pmbql and before `query-with-stages` call.

I've also added there binding of `*suppress-expression-type-check?*` to avoid warnings that come from `lib.convert/->pMBQL :dispatch-type/map`, specifically from normalization.

As the same var is already bound also in `clean-stage-schema-errors`, thanks to that previous avoids just warnings and not the clause missing effective type removal.

---

`add-types-to-fields` could then be simplified to handle pmbql only. However with that simplification in place I've started seeing lots of test failures. I suspected it is because base types are added now during the mbql -> pmbql transformation and are left in the queries during the conversion back to mbql. To mitigate that, I'm adding `:metabase.lib.query/transformation-added-base-type` key to options. If field options contain that, base type is removed during transformation back to legacy.

**Review by snoe — APPROVED** [review]

**Comment by lbrdnk** [comment]
@metabase-bot backport release-x.49.x
